File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
# variant of kadanes DP algorithm 
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        
        #1. Initialize the values 
        currentMin = float('inf') #set the min to max
        maxProfit = 0 
        
        #2. For loop 
        for price in prices:
            #a. determine the min price 
            currentMin = min(currentMin, price)
            
            #b. Calculate the profit by minimizing the min 
            currentProfit = price-currentMin
            
            #c. determine the max profit 
            maxProfit = max(currentProfit, maxProfit)
        
        return maxProfit 
        
        

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        //1. Initialize the variables (current and max)
        int currentMin = Integer.MAX_VALUE;
        int maxProfit = 0;
        
        //2. for prices 
        for(int price:prices){
            //a. process the current value 
            currentMin = Math.min(currentMin, price);
            
            //b. current profit 
            int profit = price-currentMin;
            
            //c. Max profit 
            maxProfit = Math.max(profit, maxProfit);
        }
        
        return maxProfit;
    }
}